tracer/src/Datadog.Tracer.Native/fault_tolerant_tracker.cpp

```diff
@@ -0,0 +1,172 @@
+    return _originalMethods[methodIdentifier].methodToken;
+}
+
+mdMethodDef fault_tolerant::FaultTolerantTracker::GetKickoffMethodFromInstrumentedMethod(ModuleID moduleId,
+    mdMethodDef methodId)
+{
+    std::lock_guard lock(_faultTolerantMapMutex);
+
+    const auto methodIdentifier = trace::MethodIdentifier(moduleId, methodId);
+    return _instrumentedMethods[methodIdentifier].methodToken;
+}
+
+bool fault_tolerant::FaultTolerantTracker::IsKickoffMethod(ModuleID moduleId, mdMethodDef methodId)
+{
+    std::lock_guard lock(_faultTolerantMapMutex);
+
+    const auto methodIdentifier = trace::MethodIdentifier(moduleId, methodId);
+    const auto iter = _faultTolerantMethods.find(methodIdentifier);
+    return iter != _faultTolerantMethods.end();
+}
+
+bool fault_tolerant::FaultTolerantTracker::IsOriginalMethod(ModuleID moduleId, mdMethodDef methodId)
+{
+    std::lock_guard lock(_faultTolerantMapMutex);
+
+    const auto methodIdentifier = trace::MethodIdentifier(moduleId, methodId);
+    const auto iter = _originalMethods.find(methodIdentifier);
+    return iter != _originalMethods.end();
+}
+
+bool fault_tolerant::FaultTolerantTracker::IsInstrumentedMethod(ModuleID moduleId, mdMethodDef methodId)
+{
+    std::lock_guard lock(_faultTolerantMapMutex);
+
+    const auto methodIdentifier = trace::MethodIdentifier(moduleId, methodId);
+    const auto iter = _instrumentedMethods.find(methodIdentifier);
+    return iter != _instrumentedMethods.end();
+}
+
+void fault_tolerant::FaultTolerantTracker::CacheILBodyIfEmpty(ModuleID moduleId, mdMethodDef methodId,
+    LPCBYTE pMethodBytes, ULONG methodSize)
+{
+    std::lock_guard lock(_faultTolerantMapMutex);
+
+    const auto methodIdentifier = trace::MethodIdentifier(moduleId, methodId);
+
+    if (_methodBodies.find(methodIdentifier) == _methodBodies.end())
+    {
+        // Entry does not exist, insert it
+        _methodBodies[methodIdentifier] = std::tuple(pMethodBytes, methodSize);
+    }
+}
+
+std::tuple<LPCBYTE, ULONG> fault_tolerant::FaultTolerantTracker::GetILBodyAndSize(ModuleID moduleId,
+    mdMethodDef methodId)
+{
+    std::lock_guard lock(_faultTolerantMapMutex);
+
+    const auto methodIdentifier = trace::MethodIdentifier(moduleId, methodId);
+    return _methodBodies[methodIdentifier];
+}
+
+void fault_tolerant::FaultTolerantTracker::AddSuccessfulInstrumentationVersion(
+    ModuleID moduleId, mdMethodDef methodId, const shared::WSTRING& instrumentationVersion,
+    trace::InstrumentingProducts products, std::shared_ptr<RejitHandler> rejit_handler)
+{
+    std::lock_guard lock(_successfulInstrumentationVersionsMutex);
+
+    RequestRevert(moduleId, methodId, rejit_handler);
+    RequestRejit(moduleId, methodId, rejit_handler);
+
+    const auto methodIdentifier = trace::MethodIdentifier(moduleId, methodId);
+
+    auto [iter, _] = _successfulInstrumentationVersions.emplace(methodIdentifier, std::set<shared::WSTRING>());
```

Comment:
Should the _successfulInstrumentationVersions be populated before requesting a revert and a rejit? I'm surprised this is working as-is

Reply:
Your point is totally valid and I understand why it's confusing. The fact that is working is because:
- `AddSuccessfulInstrumentationVersion` is getting called from inside the instrumentation duplicate using P/Invoke.
- `RequestRevert` and `RequestRejit` are waiting on a promise before returning.

I'll reorder the instructions so they will make more sense but the functionality will not be affected.

Reply:
Hmm I'm still a little confused. It makes sense that we enter `AddSuccessfulInstrumentationId` from the instrumentation duplicate P/Invoke and the `RequestRevert` and `RequestRejit` are waiting on a promise. However, it seems incorrect to me that we wait for the `RequestRevert` and `RequestRejit` to finish and then we update the `_successfulInstrumentationIds` map. My assumption was that the ReJIT of the kick-off method would be done by the `RequestRejit` call and that the ReJIT operation to directly instrument the method (no kick-off) required the method ID to be in the `_successfulInstrumentationIds` map. Am I missing something?

Reply:
The missing part I think is the fact that requesting rejit does not cause 'GetRejitParameters' to execute right away. It basically annotates to the CLR that the next time the method is going to get execute, I want to amend the bytecode and re-jit that method. It does not cause the method to be rejitted eagerly but it does so in on-demand fashion upon next execution/entrance

Reply:
Ah I see, now it all makes sense to me. Thank you!!